### PR TITLE
Taking a crack at new summarization API

### PR DESCRIPTION
**Pull request by keith-turner**
I finally came up with an API for ACCUMULO-4501 that I like.  I think it offers the following three things :

 * The ability to write super efficient summarizers
 * The ability to write very arbitrary summarizations
 * Gives the user explicit control over how summary info is merged.

The pull request just contains a possible API.   I am creating the PR just to review the APIs. It also contains two example summarizes.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableSummary.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  public enum Status {
+    /**
+     * Requested summary was computed for all data
+     */
+    PRESENT,
+
+    /**
+     * Requested summary were computed for some data.
+     */
+    PARTIALLY_PRESENT,
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableSummary.java]: This would be for data in memory not summarized in a file?

File: core/src/main/java/org/apache/accumulo/core/client/summary/ColumnVisibilitySummarizer.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  @Override
+  public void collect(Key k, Value v) {
+    ByteSequence cv = k.getColumnVisibilityData();
+
+    MutableLong ml = summary.get(cv);
+    if (ml == null) {
+      if (summary.size() >= MAX) {
+        ignored++;
+      } else {
+        summary.put(cv, new MutableLong(1));
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/client/summary/ColumnVisibilitySummarizer.java]: Could use LongAddr on jdk8

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableSummary.java (hunk 1 above)
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableSummary.java]: For this I was thinking of case where user requested summary data and some RFiles contained it and some did not.

**Comment by joshelser** [comment]
Took a quick glance and this looks very concise. The concrete examples are also super-duper helpful.

One thing that was in the back of my mind is how to support "bounded summaries". For example, we would want to avoid storing 1M CVs if a user had that many in a table (for some reason). Could/should this be a primitive on summaries (how to limit them)? Or, would it be some implementation detail which we could provide some common-implementation logic?

**Comment by keith-turner** [comment]
@rweeks this is an attempt to generalize what we discussed at the accumulo summit.

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableSummary.java (hunk 1 above)
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableSummary.java]: Gotcha -- thanks.

Would data resident only in memory be omitted by design? In other words, summaries are defined to only apply to RFiles and thus we don't need special API constructs to represent that condition. I think that makes sense.

Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableSummary.java]: I was only thinking of persisted data... I think summary info could be tracked for in memory maps though.  Would have to specify in interface that it should be safe to call summarize() multiple times.

**Comment by keith-turner** [comment]
> For example, we would want to avoid storing 1M CVs if a user had that many in a table (for some reason).

I think we should address this issue in some way while considering the following.

 * Fetching summaries should be relatively fast.  Gigantic summaries will stymie this goal.
 * When a users summarizer does produce a gigantic summary, it would be nice if we helped them debug it.

I am thinking one way to accomplish these goals is to store gigantic summaries, but only read summaries under a certain size.  The size of a serialized summary could be written first.  When a summary is read this size will be the first bit of info.  If the summary is over a certain size an error could be logged and that file would be treated like it had no summary.  We could also add a enum that indicates gigantic summaries were present.  Since the summary is stored, it would give the user a chance to use rfile-info to look at whats in the summary for debugging.

We also need to stress in the javadoc that summaries are intended to be small.

File: core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.Map;
+
+import org.apache.hadoop.io.Text;
+
+public interface SummaryRetriever {
+
+  /**
+   * Allows optionally setting start row before retrieving data.
+   */
+  SummaryRetriever start(Text startRow);
```
Comment (mjwall) [inline core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java]: What are you thinking about here?  What if your start row is in the middle of a split?

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java [hunk 4]
```diff
@@ -808,4 +808,11 @@ void setSamplerConfiguration(String tableName, SamplerConfiguration samplerConfi
    * @since 1.8.0
    */
   SamplerConfiguration getSamplerConfiguration(String tableName) throws TableNotFoundException, AccumuloException, AccumuloSecurityException;
+
+  /**
+   * 
+   */
+  default SummaryRetriever summaries() {
```
Comment (mjwall) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java]: Is this way you envision getting getting the summary info?  So summaries are per table, that seems right to me

File: core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java (hunk 3 above)
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java]: My assumption was that the `startRow` would essentially refer to the first Tablet to read for summaries. Since the summaries are at the file-level, you would get a summarization "greater" than what you actually specified (would be good to make sure is clear via documentation).

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java (hunk 4 above)
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java]: yeah.. I forgot to include the table name as a parameter, I intended to do that.

File: core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java (hunk 3 above)
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java]: I was thinking the first table it overlapped, summaries would be retrieved for that.

Comment (mjwall) [inline core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java]: Makes sense to me

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableSummary.java (hunk 1 above)
Comment (mjwall) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableSummary.java]: We talked a little about duplicate counts being returned, when an rfile is referenced by more than one tablet in the summarized range.  Is there an easy way to note a result included duplicate counts, or is that something we just document.

File: core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java (hunk 3 above)
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java]: I was also thinking that multiple summaries could be kept per rfile. That could work something like the following.

 * After 1,000 key values are written create a Tuple\<end row, summary\>
 * Add this tuple to an in memory list of the form List\<Tuble\<endrow, summary\>> called summaryList.  
 * If summaryList exceeds 10, then merge as merge(summaryList[0],  summaryList[1]),  merge(summaryList[2],  summaryList[3]), etc... halving the size of the list.

So basically we could easily build more granular summary info within a file that, does not exceed a fixed size, by leveraging the merge functionality of a summarizer.

File: core/src/main/java/org/apache/accumulo/core/client/summary/ColumnVisibilitySummarizer.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  private Map<ByteSequence,MutableLong> summary = new HashMap<>();
+  private long ignored = 0;
+
+  @Override
+  public String getId() {
+    return "accumulo.cvsummary";
+  }
+
+  @Override
+  public void collect(Key k, Value v) {
```
Comment (mjwall) [inline core/src/main/java/org/apache/accumulo/core/client/summary/ColumnVisibilitySummarizer.java]: Is this the function that would be used when an rfile is compacted to aggregate the counts.

File: core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java (hunk 3 above)
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java]: That's a neat idea, Keith. Sounds like it would be prudent to define an interface so that we could easily stub out different implementations on how summaries are collected/merged.

File: core/src/main/java/org/apache/accumulo/core/client/summary/ColumnVisibilitySummarizer.java (hunk 5 above)
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/summary/ColumnVisibilitySummarizer.java]: Yeah, my thinking is that every key/value added to the RFile would also be passed to this function

File: core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java (hunk 3 above)
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/SummaryRetriever.java]: @joshelser I created something called SummaryStore in ab8deb7.  It generates ~10 summaries for input data without knowing how many keys will be added beforhand.  Uses the approach I mentioned above of merging summaries.  One additional step I discovered I needed to do was double the number of key/vals per summary each time I merged.

So if I have 10 summaries each covering 1000 keys, when I merge to 5 then each of those represents 2000 keys.  So now I want start adding summaries every 2000 keys.  Next time I merge, I will start adding summaries every 4000 keys... etc.

**Comment by joshelser** [comment]
(breaking out discussion to top-level instead of rooted on a code change)

>I created something called SummaryStore in ab8deb7. It generates ~10 summaries for input data without knowing how many keys will be added beforhand. Uses the approach I mentioned above of merging summaries. One additional step I discovered I needed to do was double the number of key/vals per summary each time I merged.

> So if I have 10 summaries each covering 1000 keys, when I merge to 5 then each of those represents 2000 keys. So now I want start adding summaries every 2000 keys. Next time I merge, I will start adding summaries every 4000 keys... etc.

The [comment I left here](https://github.com/apache/accumulo/commit/ab8deb76ef96602d27072ce913ee90434bb9c246#commitcomment-20080483) has really got my brain thinking about how this should look. First, can we design a good API without defining what a summary is. Or, reworded, can we restrict what a summary is and come up with a more elegant/efficient implementation.

When it comes to `merge(Map, Map)`, I'm specifically wondering if we force all `KeyValueSummaries` to be commutative, can we manage the merging of summaries better? For example, instead of delegating to the implementation on how these two `Map`s are merged, we then get to merge them ourselves (because we know how to combine these maps).

Should a summary ever not be commutative? Does it make sense for us to allow one to not be?

**Comment by keith-turner** [comment]
> For example, instead of delegating to the implementation on how these two Maps are merged, we then get to merge them ourselves (because we know how to combine these maps).

How would we merge them?

**Comment by joshelser** [comment]
> How would we merge them?

Thinking back to the visibility example, we combine the values for common keys between the maps and retain `N` entries with the largest values. Perhaps I didn't do a good job of posing the question explicitly: is it possible to implement this merge generically? Are there cases which we could/should support which would require some specialized knowledge or understanding to merge?

**Comment by keith-turner** [comment]
> Perhaps I didn't do a good job of posing the question explicitly: is it possible to implement this merge generically? Are there cases which we could/should support which would require some specialized knowledge or understanding to merge?

Yesterday I toyed around with changing `merge(Map, Map)` to `merge(String, long, long)` and this worked beautifully for the `ExampleSummarizer`.  Its merge method became something like the following :

```java
  public void merge(String summaryKey, long v1, long v2){
     switch(summaryKey){
        case "deletes":
        case "total":
           return v1 + v2;
        case "minStamp":
          return Math.min(v1, v2);
       case "maxStamp":
          return Math.max(v1,v2);
     }
  }
```

Then I went to change `ColumnVisibilitySummarizer` to the new interface and found it would not work at all.  Maybe would could offer a an Abstract subclass that has an abstract `merge(String, long, long)` method?  I might try this just to see what it looks like.

**Comment by keith-turner** [comment]
I have been thinking about the case of a buggy summarizer and the possibility of adding an explicit version numbers to serializers.  I decided that it adds a lot of complications and there is an easier way for users to handle it.  Assume I discover that `ExampleSummarizer` is generating data for `deletes` incorrectly.  A simple way to handle this is to push out a new version of the code that generates a new key like `deletes.v2` and merge `deletes` and `deletes.v2` separately.

**Comment by keith-turner** [comment]
Superseded by #224
